From 00e94f40d08470ba87adc52177997966ea2b069d Mon Sep 17 00:00:00 2001
From: Aloka Dixit <quic_alokad@quicinc.com>
Date: Mon, 24 Jan 2022 10:53:22 -0800
Subject: [PATCH] ath12k: configure WPA and RSNE parameters for nontransmitting
 interface

Set vht_cap, rsnie_present and wpaie_present fields for the
non-transmitting interfaces when MBSSID is enabled.

Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 90 ++++++++++++++++++++++++++-
 1 file changed, 89 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index aea8429..7f8a7b4 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1317,6 +1317,79 @@ static void ath12k_mac_setup_bcn_tmpl_vif_params(struct ath12k_vif *arvif,
 		arvif->wpaie_present = true;
 }
 
+static bool ath12k_mac_setup_bcn_tmpl_nontx_vif_params(struct ath12k_vif *tx_arvif,
+						       struct ath12k_vif *arvif,
+						       struct sk_buff *bcn)
+{
+	struct ieee80211_mgmt *mgmt;
+	const u8 *ies, *profile, *next_profile;
+	int ies_len;
+
+	if (arvif == tx_arvif)
+		return true;
+
+	arvif->rsnie_present = tx_arvif->rsnie_present;
+
+	ies = bcn->data + ieee80211_get_hdrlen_from_skb(bcn);
+	ies += sizeof(mgmt->u.beacon);
+	ies_len = skb_tail_pointer(bcn) - ies;
+
+	ies = cfg80211_find_ie(WLAN_EID_MULTIPLE_BSSID, ies, ies_len);
+
+	while (ies) {
+		u8 mbssid_len;
+
+		ies_len -= (2 + ies[1]);
+		mbssid_len = ies[1] - 1;
+		profile = &ies[3];
+
+		while (mbssid_len) {
+			u8 profile_len, nie_len, *nie;
+
+			profile_len = profile[1];
+			next_profile = profile + (2 + profile_len);
+			mbssid_len -= (2 + profile_len);
+
+			profile += 2;
+			profile_len -= (2 + profile[1]);
+			profile += (2 + profile[1]); /* nontx capabilities */
+			profile_len -= (2 + profile[1]);
+			profile += (2 + profile[1]); /* SSID */
+			if (profile[2] == arvif->vif->bss_conf.bssid_index) {
+				profile_len -= 5;
+				profile = profile + 5;
+
+				if (cfg80211_find_ie(WLAN_EID_RSN, profile,
+						     profile_len))
+					arvif->rsnie_present = true;
+				else if (tx_arvif->rsnie_present) {
+					nie = cfg80211_find_ext_ie(WLAN_EID_EXT_NON_INHERITANCE,
+								   profile,
+								   profile_len);
+					if (nie) {
+						int i;
+
+						nie_len = nie[1];
+						nie += 2;
+						for (i = 0; i < nie_len; i++) {
+							if (nie[i] ==
+								WLAN_EID_RSN) {
+								arvif->rsnie_present = false;
+								break;
+							}
+						}
+					}
+				}
+				return true;
+			}
+			profile = next_profile;
+		}
+		ies = cfg80211_find_ie(WLAN_EID_MULTIPLE_BSSID, profile,
+				       ies_len);
+	}
+
+	return false;
+}
 
 static int __ath12k_mac_setup_bcn_tmpl(struct ath12k_vif *arvif,
 				       struct sk_buff *bcn,
@@ -1344,6 +1417,7 @@ static int ath12k_mac_setup_bcn_tmpl_ema(struct ath12k_vif *arvif)
 	struct ieee80211_ema_beacons *beacons;
 	u8 i = 0;
 	int ret = 0;
+	bool found_vdev = false;
 
 	if (!arvif->vif->mbssid_tx_vif)
 		return -1;
@@ -1357,11 +1431,21 @@ static int ath12k_mac_setup_bcn_tmpl_ema(struct ath12k_vif *arvif)
 		return -EPERM;
 	}
 
-	if (tx_arvif == arvif)
+	if (tx_arvif == arvif) {
 		ath12k_mac_setup_bcn_tmpl_vif_params(tx_arvif,
 						     beacons->bcn[0].skb);
+		found_vdev = true;
+	} else {
+		arvif->wpaie_present= tx_arvif->wpaie_present;
+	}
 
 	for (i = 0; i < beacons->cnt; i++) {
+		if (found_vdev == false)
+			found_vdev =
+				ath12k_mac_setup_bcn_tmpl_nontx_vif_params(tx_arvif,
+									   arvif,
+									   beacons->bcn[i].skb);
+
 		ret = __ath12k_mac_setup_bcn_tmpl(tx_arvif, beacons->bcn[i].skb,
 						  beacons->bcn[i].offs,
 						  i, beacons->cnt);
@@ -1397,6 +1481,10 @@ static int ath12k_mac_setup_bcn_tmpl_non_ema(struct ath12k_vif *arvif)
 
 	if (tx_arvif == arvif)
 		ath12k_mac_setup_bcn_tmpl_vif_params(tx_arvif, bcn);
+	else
+		(void) ath12k_mac_setup_bcn_tmpl_nontx_vif_params(tx_arvif,
+								  arvif,
+								  bcn);
 
 	ret = __ath12k_mac_setup_bcn_tmpl(tx_arvif, bcn, offs, 0, 0);
 	if (ret)
-- 
2.17.1

